src/frames.ts

```diff
@@ -386,39 +402,53 @@ export class Frame {
       }
       url = helper.completeUserURL(url);
 
-      const frameTask = new FrameTask(this, progress);
-      const sameDocumentPromise = frameTask.waitForSameDocumentNavigation();
-      const navigateResult = await this._page._delegate.navigateFrame(this, url, referer).catch(e => {
-        // Do not leave sameDocumentPromise unhandled.
-        sameDocumentPromise.catch(e => {});
-        throw e;
-      });
-      let request: network.Request | undefined;
+      const sameDocument = helper.waitForEvent(progress, this, kNavigationEvent, (event: NavigationEvent) => !event.documentInfo);
+      const navigateResult = await this._page._delegate.navigateFrame(this, url, referer);
+
+      let event: NavigationEvent;
       if (navigateResult.newDocumentId) {
-        // Do not leave sameDocumentPromise unhandled.
-        sameDocumentPromise.catch(e => {});
-        request = await frameTask.waitForSpecificDocument(navigateResult.newDocumentId);
+        sameDocument.dispose();
+        event = await helper.waitForEvent(progress, this, kNavigationEvent, (event: NavigationEvent) => {
+          // We are interested either in this specific document, or any other document that
+          // did commit and replaced the exepcted document.
+          return event.documentInfo && (event.documentInfo.documentId === navigateResult.newDocumentId || !event.error);
+        }).promise;
+
+        if (event.documentInfo!.documentId !== navigateResult.newDocumentId)
+          throw new Error('Navigation interrupted by another one');
```

Comment:
Let's get ourselves a test

Reply:
Added a test, but I cannot trigger this particular error - we get cancellation instead. I'll leave a comment here.